### DISCUSSION TITLE
Successfully changing the value of an input is still too difficult, with a React app at least.  Should we put best practice guidelines for it in the docs?

### DISCUSSION THREAD

**Discussion by jedwards1211** [webdriverio/webdriverio#7650] [General]
If I recall correctly:

* `setValue` and `clearValue` behavior is determined by the webdriver spec, not this project
* `setValue` doesn't clear the value first, it just types the given text one keystroke at a time
* `clearValue` doesn't work with React controlled inputs

Because of these onerous circumstances, my workaround has evolved to this ridiculous point:

```js
export default async function setValue(
  target: string,
  value: any,
  {
    killAutocomplete = true,
    pause,
  }: { killAutocomplete?: boolean, pause?: number } = {}
) {
  value = String(value)
  const el = await browser.$(target)
  let remainingTries = 5
  while (remainingTries-- > 0 && (await el.getValue()) !== value) {
    await el.click()
    let keys = ['End']
    for (let i = 0; i < (await el.getValue()).length; i++) {
      keys.push('Backspace')
    }
    await browser.keys(keys)
    if (pause) await delay(pause)
    await el.setValue(value)
    if (killAutocomplete) {
      // kill autocomplete popup...otherwise it will stomp on the value we entered
      await browser.keys(['Escape'])
    }
  }
  expect(
    await el.getValue(),
    `failed to set the value of <${target}> to ${JSON.stringify(value)}`
  ).to.equal(value)
}
```

This feels like an all-around toxic situation, and I imagine it's daunting for people who are just getting started with webdriver.io, because it took me a long time to figure out these limitations and workarounds.

First off, is there anything I could be doing better?
And secondly, if webdriver.io isn't going to provide a utility function with all of these workarounds, should we at least put it in the documentation, so that newcomers don't have to go through the painful discovery process I did?

**Comment by christian-bromann** [comment]
> * `setValue` doesn't clear the value first, it just types the given text one keystroke at a time

Not quite, `setValue` is the combination out of `clearValue` and `addValue`.

Can you provide a fiddle example where we can test some workarounds? I am happy to improve the `addValue` command to better work with these framework inputs.

I would also make sense to create an issue in the React project to make the team aware that their input elements aren't compatible with WebDriver based automation tools.

**Reply by jedwards1211** [reply]
I'll have to hunt for what I saw, but IIRC they're aware of it and had brought it up with WebDriver folks but WebDriver folks unfortunately weren't open to making a clearValue type command that realistically simulates user interaction with the input.

Haven't people brought up issues with React before in this repo?  Surely I'm not the only one dealing with this...

**Reply by jedwards1211** [reply]
The problem is with `clearValue`, not `addValue`

**Reply by jedwards1211** [reply]
Okay I did some research, and it's a problem with the WebDriver specification for `clearValue`.

`setValue` isn't a WebDriver command itself, just a `webdriver.io` utility, right?  Hopefully I can convince you that it would be best not to rely on `clearValue` in your `setValue` implementation...If nothing else it would be good to give people a heads up about this in the docs.

Here's the spec for `clearValue`:

> To clear a resettable element:
> 
> Let empty be the result of the first matching condition:
> 
> element is an input element whose type attribute is in the File Upload state
> True if the list of selected files has a length of 0, and false otherwise.
> Otherwise
> True if its value IDL attribute is an empty string, and false otherwise.
> If element is a candidate for constraint validation it satisfies its constraints, and empty is true, abort these substeps.
> 
> Invoke the focusing steps for element.
> 
> Invoke the clear algorithm for element.
> 
> **Invoke the unfocusing steps for the element.**
> 

> The clear algorithm for input elements is to set the dirty value flag and dirty checkedness flag back to false, **set the value of the element to an empty string**, set the checkedness of the element to true if the element has a checked content attribute and false if it does not, empty the list of selected files, and then invoke the value sanitization algorithm, if the type attribute’s current state defines one.

There are two problems with this:
1. It doesn't trigger an `input` event on the element
2. It blurs it (meaning `clearValue` followed by `addValue` is an unrealistic simulation of user interaction)

Unfortunately, some React form libraries have a blur handler that ends up restoring the value from what it had in internal state, since it never got an `input` event for the cleared value.

I haven't set up webdriver.io code yet but if you go to [this sandbox](https://codesandbox.io/s/beautiful-poincare-6zyue?file=/src/App.js), you can see the problem if you open the console at bottom right (the console codesandbox provides, not your browser dev tools) and type `theInput.value = ''`: it clears the event but no DOM or React event gets logged.  Compare this to if you select all in the input and hit backspace; an event gets logged.

A more reasonable simulation of user interaction for setting the value would be:
1. Focus the input
2. Select all text
3. Type new value
4. blur the input (or at least press escape if a browser autofill popup appears)

Here's the [original Selenium issue](https://github.com/SeleniumHQ/selenium/issues/6741) and the [chromedriver issue closed as wontfix](https://bugs.chromium.org/p/chromedriver/issues/detail?id=2702&q=&sort=-id&colspec=ID%20Status%20Pri%20Owner%20Summary) about `clearValue`.

**Reply by christian-bromann** [reply]
> `setValue` isn't a WebDriver command itself, just a `webdriver.io` utility, right?

Correct. It is the same as `addValue` with the difference that it calls `clearValue` before adding a value.

> There are two problems with this:

Mind raising an issue in the [spec](https://github.com/w3c/webdriver)?

> Unfortunately, some React form libraries have a blur handler that ends up restoring the value from what it had in internal state, since it never got an `input` event for the cleared value.

We could also see this as a bug in React. I wonder if the team knows about it. Are there any issues logged in the React project about this?

> 4\. blur the input (or at least press escape if a browser autofill popup appears)

I believe this is what `clearValue` does, see the comment in the Chromedriver bug:

>  In particular, https://www.w3.org/TR/2018/REC-webdriver1-20180605/#dfn-clear-a-resettable-element step 5 requires ChromeDriver to move the focus away from the element that has just been cleared. This jscolor control has a blur event handler that "repairs" the value inside the input when it loses focus, and unfortunately this interferes with what Clear does.

I am open for suggestion how we can make this more stable and work with React inputs. Be aware that WebdriverIO doesn't know the framework in which the application under test is build in. Every change we make to `setValue` will also be applied to applications that are not React based.

**Reply by jedwards1211** [reply]
I totally understand about being framework agnostic.  But this can't properly be called a bug in React because it's just a consequence of these two things:
1. Using the one-way data flow encouraged by React where you control the value of inputs and respond to events instead of always reading the `value` property of inputs
2. The WebDriver clear algorithm not firing an event to reflect the state change it's made (I need to make sure that it's really not firing an event when I run `clearValue` from a test, but pretty sure that's what's happening)

If you can't rely on events to stay informed of input changes, your only options are to poll, or at least just read the DOM attributes on submit, and I don't think anyone intends for those to be the only options.  So I only consider the WebDriver spec to be buggy here.  Yeah I'll raise an issue in the spec 👍

It's not a bug with React, just an architectural pattern designed to reduce the confusion and headaches that come from the old two-way bindings style of working with the DOM.  Maybe it could be considered a bug with the form libraries, but I need to look into what they're doing with blur and why.

**Reply by jedwards1211** [reply]
And regardless of frameworks, I hope you'll agree that it's not ideal for `setValue` to cause the input to get blurred between clearing it and typing in the new value.  Since that's significantly different from user interaction, and code can respond to blur events, it could trigger behavior that devs don't expect

**Reply by jedwards1211** [reply]
Okay with `react-final-form` at least, I was wrong about the blur handler doing anything.  Rather, here is the steps of what happens:
- clear algorithm sets the input value to `''` but doesn't trigger event, so the old value still exists in `final-form` internal state
- a `react-final-form` focus handler updates internal state (it tracks field focus for a bunch of reasons)
- the state update triggers a rerender of the React components
- the components render with the old value from `final-form` internal state
- React reconciles the virtual DOM with the real DOM, restoring the old value

`react-final-form` also has a `formatOnBlur` option; if that were used it would call the format function with the old value from `final-form` internal state.

```
it(`react-final-form and setValue`, async function () {
  await browser.url('https://6zyue.csb.app/')
  await browser.$('#formInput').setValue('test')
  expect(await browser.$('#formInput').getValue()).to.equal('test')
})
```

**Reply by jedwards1211** [reply]
Okay here's the issue I filed in the WebDriver spec project: https://github.com/w3c/webdriver/issues/1630

**Comment by jedwards1211** [comment]
To reiterate, I think we should concretely specify the behavior of `setValue` as one of the following options, and only use `clearValue` if it's compatible with the specified behavior:

### `setValue` Option 1
- focus the input
- select all text (seems like we have to do this by sending a browser script?)
- send backspace key clear the text
- **without blurring the input**, type the new value
- blur the input

### `setValue` Option 2 (less preferable)
- focus the input
- **without blurring the input**, set the input's value **in a way that triggers an event where e.target.value will be the new value**
- blur the input

**Comment by christian-bromann** [comment]
@jedwards1211 Thanks for the feedback, let's see what the spec writers suggest in this case.

**Reply by jedwards1211** [reply]
@christian-bromann turns out there was a dupe of my WebDriver issue.  In that old thread, one of the contributors explicitly states that element clear isn't intended to be an emulation of user behavior: https://github.com/w3c/webdriver/issues/445#issuecomment-273187301

Given that, I think webdriver.io should implement its own `clearValue` method that does emulate user behavior

**Comment by jedwards1211** [comment]
@christian-bromann I had another thought about this...

A real user changing an input value would do something like
- Select all text with Ctrl/Command-A
- Select all text with the mouse cursor
- Move the cursor to the end and then Ctrl-Backspace
- etc.

Some application code could have key or selection listeners that react to these actions.  Imagine that there is a bug in one of those event listeners.  If the test code doesn't trigger those events, the tests would pass, but real users would experience those bugs.

But if test code realistically emulates at least one of the things a real user would do, it would be more likely to catch those bugs.  This would be an added benefit on top of not running into problems with React controlled inputs.

So to me, philosophically, even if webdriver implementations are fixed to fire the change event during the unfocusing steps, it's not the best way to change an input value for testing anyway.  What do you think?

**Reply by christian-bromann** [reply]
I've never seen an example where a form has certain listeners on how to text in input elements is cleared. However I agree that commands should imitate the user _as close as possible_. What would you suggest is the best way forward? Have a parameter to `setValue` that defines how someone clears the value?

**Reply by jedwards1211** [reply]
Not that they would have listeners like that for when the text is cleared.  But I assume things like a 3-input phone number field where the cursor jumps to the next field when you finish filling in the previous one are usually implemented with selection listeners.  If logic like that is flawed it could interfere when a real user tries to clear the input or replace the value.

Having a parameter sounds like a good idea to me.  Of course, an additional option would be to select all and then type the new value without even clearing first (unless the new value is the empty string)

**Reply by jedwards1211** [reply]
I do wish WebDriver had a select all command, because we'd have to look at the user agent to decide whether to send Control A or Meta A, and I'm not sure if we can completely trust that?
On the other hand we could do `browser.execute(() => document.activeElement?.select())` but that wouldn't trigger the key events.
I'm also not sure if sending the End key followed by enough backspaces would work on every platform.  If it's hard to get something reliable enough, I understand...but in that case I think we should propose new commands in the WebDriver spec.

**Comment by fdomfeh** [comment]
The following code I came up with should do the magic for everyone

```
export async function clearText(selector: WebdriverIO.Element) {
    let valueLength = (await selector.getValue()).length
    let backSpaces = new Array(valueLength).fill('Backspace')
    await selector.setValue(backSpaces);
```
};

Above method will clear the input value and now you can use the  `setValue(" ")` to set a new value

**Reply by jedwards1211** [reply]
@christian-bromann is `setValue` guaranteed to position the cursor at the end of the input before sending the keys?

**Reply by christian-bromann** [reply]
I am not sure if the WebDriver protocol explicitly defines this.
